### ISSUE TITLE
Minitest::Assertions#diff doesn't work reliably

### ISSUE THREAD

**Issue by cboos** [seattlerb/minitest#585]
I just noticed an issue with Minitest after switching from Ruby 1.9.3 (x86) to Ruby 2.2.2 (x64), on Windows.

For an expected failing test, instead of the nice diff output I used to get with 1.9.3, I now see the awfully confusing message:
```
No visible difference in the String#inspect output.
You should look at the implementation of #== on String or its members.
```
First I was side-tracked by #476, but my problem is really not about encodings.

The comparison within `Minitest::Assertions#diff` fails because at the time `diff.exe -u` is executed on the temporary files, their content is *empty*. Indeed, the doc for `flush` (http://ruby-doc.org/core-2.2.3/IO.html#method-i-flush) specifies that:
```
 (note that this is Ruby internal buffering only; the OS may buffer the data as well).
```
So far this wasn't a problem and the data ended up on disk anyway before diff.exe started to read it, but maybe Ruby 2.2.2 is faster at spawning external processes.

In any case, calling `fsync` after each `fflush` does the trick and diff works again.

```diff
--- ./ruby/gems/2.2.0/gems/minitest-5.8.1/lib/minitest/assertions.rb.orig       Fri Oct 09 14:30:43 2015
+++ ./ruby/gems/2.2.0/gems/minitest-5.8.1/lib/minitest/assertions.rb    Fri Oct 09 14:31:38 2015
@@ -73,10 +73,18 @@
       Tempfile.open("expect") do |a|
         a.puts expect
         a.flush
+        begin
+          a.fsync
+        rescue NotImplementedError
+        end

         Tempfile.open("butwas") do |b|
           b.puts butwas
           b.flush
+          begin
+            b.fsync
+          rescue NotImplementedError
+          end

           result = `#{Minitest::Assertions.diff} #{a.path} #{b.path}`
           result.sub!(/^\-\-\- .+/, "--- expected")
```

(should I create a pull request for this simple fix or is this bug report enough?)

**Comment by zenspider** [comment]
In four years, this is the first that this type of thing has been reported. I don't use windows so I can't test your patch. I'm hesitant to believe that this problem exists across windows + ruby. Whether fsync occurs or not _shouldn't_ matter as much because `diff` should be accessing via the same OS buffers (just like any other OS).

What diff are you using? Are you on any particularly different disk or OS when you're running? IE, Is there anything special going on?

**Comment by cboos** [comment]
Well, I can try to dig further in the Ruby code and the Windows API if you want, but my reading of *this is Ruby internal buffering only; the OS may buffer the data as well* is that the *process* itself might still buffer it, with no guarantee that *another* process could see that content until it actually reaches the disk. At least, this interpretation is consistent with what I observed. No, it's even worse than that, now I remember: I also tried to *reopen* the temp file with a `File.read` just after the `.flush`, and it returned an empty string...

**Comment by cboos** [comment]
Sorry, I didn't remember correctly all the details. Let's go through it once more together if you don't mind.

With `spec.rb`:
```ruby
gem 'minitest'
require 'minitest/autorun'
require 'minitest/spec'

describe 'minitest-issue-585' do
  it 'should show me a nice diff output' do
    value(%(
Let's have a few lines here
...
)).must_equal (%(
Let's have a few more lines here
...
...
))
  end
end
```

I get the following without the `fsync` patch:
```
 $ ruby spec.rb
Run options: --seed 46779

# Running:

F

Finished in 0.045647s, 21.9073 runs/s, 21.9073 assertions/s.

  1) Failure:
minitest-issue-585#test_0001_should show me a nice diff output [spec.rb:8]:
No visible difference in the String#inspect output.
You should look at the implementation of #== on String or its members.
"
Let's have a few more lines here
...
...
"

1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

With the `fsync` patch:
```
ruby spec.rb --seed 46779
Run options: --seed 46779

# Running:

F

Finished in 0.073177s, 13.6654 runs/s, 13.6654 assertions/s.

  1) Failure:
minitest-issue-585#test_0001_should show me a nice diff output [spec.rb:8]:
--- expected
+++ actual
@@ -1,5 +1,4 @@
 "
-Let's have a few more lines here
-...
+Let's have a few lines here
 ...
 "


1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

And now with this patch, doing the `File.read`:
```diff
assertions.rb.orig assertions.rb
--- assertions.rb.orig  Fri Oct 09 14:30:43 2015
+++ assertions.rb       Fri Oct 23 20:25:10 2015
@@ -78,6 +78,7 @@
           b.puts butwas
           b.flush

+          p [:a, File.read(a.path), :b, File.read(b.path)]
           result = `#{Minitest::Assertions.diff} #{a.path} #{b.path}`
           result.sub!(/^\-\-\- .+/, "--- expected")
           result.sub!(/^\+\+\+ .+/, "+++ actual")
```

I get this:
```
ruby spec.rb --seed 46779
Run options: --seed 46779

# Running:

[:a, "\"\nLet's have a few more lines here\n...\n...\n\"\n", :b, "\"\nLet's have a few lines here\n...\n\"\n"]
F

Finished in 0.053808s, 18.5845 runs/s, 18.5845 assertions/s.

  1) Failure:
minitest-issue-585#test_0001_should show me a nice diff output [spec.rb:8]:
--- expected
+++ actual
@@ -1,5 +1,4 @@
 "
-Let's have a few more lines here
-...
+Let's have a few lines here
 ...
 "


1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```
So I misremembered that one, if I do the `File.read` from within the same process, it indeed works and even the follow-up `diff.exe`.

Even doing an `ls` instead of a `diff` seems to work(?):
```
ruby spec.rb --seed 46779
Run options: --seed 46779

# Running:

F

Finished in 0.079885s, 12.5181 runs/s, 12.5181 assertions/s.

  1) Failure:
minitest-issue-585#test_0001_should show me a nice diff output [spec.rb:8]:
-rw-r--r--    1 cboos    Administ       40 Oct 23  2015 C:/Users/cboos/AppData/Local/Temp/butwas20151023-19536-1r1lecd -rw-r--r--    1 cboos    Administ       50 Oct 23  2015 C:/Users/cboos/AppData/Local/Temp/expect20151023-19536-iie29
...
```

It also works if `diff --binary` (?), but definitely not with `diff -a` (text mode) which seems to be the default behavior. Most likely (one more) Windows oddity about this special text mode...

So a better fix would perhaps be:
```diff
--- assertions.rb.orig  Fri Oct 09 14:30:43 2015
+++ assertions.rb       Fri Oct 23 20:39:54 2015
@@ -27,7 +27,7 @@
     def self.diff
       @diff = if (RbConfig::CONFIG["host_os"] =~ /mswin|mingw/ &&
                   system("diff.exe", __FILE__, __FILE__)) then
-                "diff.exe -u"
+                "diff.exe --binary -u"
               elsif Minitest::Test.maglev? then
                 "diff -u"
               elsif system("gdiff", __FILE__, __FILE__)
```
Which is also enough to get it to work.

**Comment by cboos** [comment]
Ha! With another diff.exe, no patch needed (even when I force a `diff -a`, it still works). 

The diff.exe that works is the one which come from my Msys-Git installation, and that diff is linked to kernel32.dll and msys-1.0.dll. The version says *diff - GNU diffutils version 2.7*.

The one which doesn't work is the one which was bundled with the Vim 74 "[cream](http://sourceforge.net/projects/cream/files/Vim/7.4.827/gvim-7-4-827.exe/download)" distribution for Windows , and it is linked to kernel32.dll and msvcrt.dll only. It also says *diff - GNU diffutils version 2.7*.

I also tried a 3rd diff, the one from the [GnuWin 32](http://gnuwin32.sourceforge.net/packages/diffutils.htm) project, and that one works too.

The weird thing is that this msvcrt.dll version of diff.exe also works fine with Minitest-5.8.1 for Ruby 1.9.3, 2.1.5 (32 and 64 bits), but not for the 2.2 (64 bits) one (*ruby 2.2.2p95 (2015-04-13 revision 50295) [x64-mingw32]*), and also not for a 2.2.1 (32 bits) version (*ruby 2.2.1p85 (2015-02-26 revision 49769) [i386-mingw32]*). All these Ruby versions come from rubyinstaller.org.

Bottom line, something must have changed in these versions of Ruby (or with the way they're built) which exposes a bug or strange behavior in some GNU diff builds (e.g.when the latter is built directly for the win32 API). I think now it's probably not worth changing anything in Minitest (as I'm not sure if all version of diff on Windows support the `--binary` flag). If someone else experiences this problem, they'll probably find this issue and see that switching to the msys diff (or the GnuWin32 one) will fix the issue.

**Comment by cboos** [comment]
After some additional research, I found that the first thing diff 2.7 does with its input files is to get their size via a call to `stat`. If it gets 0, then it will not be able to read the files, even if there would be something to read. And this misleading value of 0 is what the UnxUtils build gets, as `stat` in MSVCRT.dll will call `FindFirstFileEx` which obtains the file size info efficiently but unreliably from the **directory entry**, not from the file itself. On NTFS, the file size info in the directory entry gets only updated at special times, like when the file is closed. See for example the detailed explanation in [Why is the file size reported incorrectly for files that are still being written to?](http://blogs.msdn.com/b/oldnewthing/archive/2011/12/26/10251026.aspx).

Another situation in which the directory entry gets updated with the correct file size is after a `fsync`, (`fsync` is aliased to `_commit`, which does a call to `FlushFileBuffers`), as explained in [CreateFile - Caching Behavior](https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx#caching_behavior): *To ensure that the metadata is flushed to disk, use the FlushFileBuffers function*.

This explains why this particular diff.exe linked with the msvcrt doesn't work when called from assertions.rb, as both temp files are still opened at that time: `stat` will report a file size of 0 and diff won't read the files. By contrast, in the the msys version of diff.exe `stat` works differently and gets the file size directly from the file (calling `fstat` which calls `GetFileSize`), so it will work even when the files are still opened and not `fsync`ed.

Now the difference between the earlier Ruby versions which don't exhibit this problem and the more recent ones which do, is precisely that Ruby's **`IO.flush` no longer does a fsync** on Windows. There was a `fsync` call in `IO.flush` which was removed for performance reasons in r45254, according to [Ruby Bug #9153](https://bugs.ruby-lang.org/issues/9153). The description of this bug gives a very clear presentation of the whole story, and why this fsync was introduced in the first place, in order to work around similar problems with wrong file size (see also [Ruby Bug #776](http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/20043))!

So while I think that removing `fsync` in general is probably a good thing performance wise, in "our" specific case, as we know that some versions of the diff tool we're calling will misbehave if we're not doing the `fsync`, it can be worthwhile to add it back. 

Note that the other workaround I found (pass the `--binary` flag) is not robust enough. After having debugged diff.exe, I see that its only effect is to prevent diff to throw away the "probe" read it did to see if the input was binary or not, with the file opened in binary mode. So it will only work if the difference can be found within this initial buffer. Otherwise if the difference comes after byte 8192, it will not be seen and we'll have the *No visible difference in the String#inspect output.* result again.

**Comment by cfis** [comment]
Just a heads up - I also see this issue also on windows. Environment:

Windows 10
ruby 2.3.1p112 (2016-04-26 revision 54768) [x64-mingw32]
msys2
diff (GNU diffutils) 3.3

The fix of adding the two fsync calls to the #diff method fixes it here.
